### DISCUSSION TITLE
Having trouble loading assets in a notebook

### DISCUSSION THREAD

**Discussion by seeess1** [catalyst-cooperative/pudl#3772] [Help Me!]
Hi! I tackled my first PUDL ticket [here](https://github.com/catalyst-cooperative/pudl/pull/3670) (relevant issue [here](https://github.com/catalyst-cooperative/pudl/issues/3519)). For the most part things went pretty smooth. But I was never able to actually load the asset locally into a notebook with:

```py
from dagster import AssetKey
from pudl.etl import defs

demand_forecasts = defs.load_asset_value(AssetKey(<name of asset>))
```

So I had to do a sorta clunky workaround (you can see what I mean in `notebooks/work-in-progress/ferc714-core.ipynb` in my [merged PR](https://github.com/catalyst-cooperative/pudl/pull/3670/files)). Doing that unblocked me but I came back afterward to try and figure out how to load assets using the logic posted above. I ended up trying a few things including rebuilding my pudl environment from scratch (which had its own complications) but nothing worked. Am hoping to get some advice/potential solutions here on how I can set up my machine so I can easily load assets into notebooks/scripts for local development.

Also, for context, I have all the raw/SQLite PUDL data from the ETL in a directory called `pudl-data` that sits next to the `pudl` repo on my machine. So my `dagster_home` lives here:

```
/Users/sam/Documents/pudl-data/dagster_home
```

And `pudl_output` lives here:

```
/Users/sam/Documents/pudl-data/pudl_output
```

And my `pudl` repo lives here:

```
/Users/sam/Documents/pudl
```

**Comment by zaneselvans** [comment]
Okay, some questions to debug with...

* What operating system are you working in?
* In your newly recreated `pudl-dev` environment are you able to run `dagster` at the commnd line, and materialize assets using the Dagster UI / web interface?
* Have you already materialized the asset (through the Dagster UI) that you're trying to load?
* Have you tried running that very short 3-line script above to load an asset from within a python prompt in the terminal with the `pudl-dev` environment activated?
* Can you past the error messages that you're getting when you try to run the above script in your notebook?
* How are you running the Jupyter Notebook in which you're unable to load the asset? Are you sure it's running within the `pudl-dev` environment? How can you check?
* Can you read your `$DAGSTER_HOME` and `$PUDL_OUTPUT` environment variables from within the notebook? For example:

```py
import os
print(os.environ["DAGSTER_HOME"])
# '/Users/zane/code/catalyst/dagster-home'
print(os.environ["PUDL_OUTPUT"])
# '/Users/zane/code/catalyst/pudl-output'
```

Similarly, from within the notebook are you able to import libraries that should be available in `pudl-dev`?

```py
import dagster
import pudl
```

**Reply by zaneselvans** [reply]
Also, if you list the files in your `$DAGSTER_HOME` directory, is the parquet file associated with the asset you're trying to load present?  E.g. if you're doing:

```py
from dagster import AssetKey
from pudl.etl import defs
df = defs.load_asset_value(AssetKey("out_eia__yearly_plants"))
```

What do you find if you do:
```console
ls -l $PUDL_OUTPUT/parquet/out_eia__yearly_plants.parquet
# -rw-r--r-- 1 zane staff 2.9M 2024-07-29 20:59 /Users/zane/code/catalyst/pudl-output/parquet/out_eia__yearly_plants.parquet
```

**Reply by seeess1** [reply]
> Okay, some questions to debug with...
> 
> * What operating system are you working in?
> * In your newly recreated `pudl-dev` environment are you able to run `dagster` at the commnd line, and materialize assets using the Dagster UI / web interface?
> * Have you already materialized the asset (through the Dagster UI) that you're trying to load?
> * Have you tried running that very short 3-line script above to load an asset from within a python prompt in the terminal with the `pudl-dev` environment activated?
> * Can you past the error messages that you're getting when you try to run the above script in your notebook?
> * How are you running the Jupyter Notebook in which you're unable to load the asset? Are you sure it's running within the `pudl-dev` environment? How can you check?
> * Can you read your `$DAGSTER_HOME` and `$PUDL_OUTPUT` environment variables from within the notebook? For example:
> 
> ```python
> import os
> print(os.environ["DAGSTER_HOME"])
> # '/Users/zane/code/catalyst/dagster-home'
> print(os.environ["PUDL_OUTPUT"])
> # '/Users/zane/code/catalyst/pudl-output'
> ```
> 
> Similarly, from within the notebook are you able to import libraries that should be available in `pudl-dev`?
> 
> ```python
> import dagster
> import pudl
> ```

---

Thanks for providing these steps. Here's where I'm at (added ❌ and ✅ to make it more clear what's working for me right now and what isn't):

### 1 - OS Version
I'm on Mac OS (Ventura 13.6.7).

### 2 - Running Dagster and materializations ❌
I'm able to run `dagster-webserver -m pudl.etl -m pudl.ferc_to_sqlite` from the command line and go to the Dagster interface. I'm able to materialize raw assets but core assets are failing. When I tried just now to materialize [the core asset I added](https://github.com/catalyst-cooperative/pudl/pull/3670/files) in my PR (`core_ferc714__yearly_planning_area_demand_forecast`), I got the following error:

```
RuntimeError: Database schema has changed, run `alembic revision --autogenerate -m 'relevant message' && alembic upgrade head`.
```

So I ran that command but got a different error:

```
ERROR [alembic.util.messaging] Can't locate revision identified by '2c52a938f5cc'
```

I know [the docs](https://catalystcoop-pudl.readthedocs.io/en/nightly/dev/run_the_etl.html) have information about this situation but the revision ID in my error message (`2c52a938f5cc`) has no hits anywhere in my codebase. So I can't follow the instruction to check out the git branch that with the errant migration ID. I'm also on `main` and have the latest changes pulled in from the `pudl` repo so I'm not really sure where to look.

### 3 - Previous materializations ✅
I had previously materialized the `core_ferc714__yearly_planning_area_demand_forecast` asset when I was working on this PR:

![image](https://github.com/user-attachments/assets/2583b450-2f8f-4cba-92ce-8385de845f9f)

### 4 - Importing env vars ✅
I can run the following as a Python script or Jupyter notebook (slight qualification on the notebook part described further down):

```
import os

print(os.environ["DAGSTER_HOME"])
print(os.environ["PUDL_OUTPUT"])
```

Output:

```
/Users/sam/Documents/pudl-data/dagster_home
/Users/sam/Documents/pudl-data/pudl_output
```

I was having an issue before where running a notebook was erroring out. But it [looks like](https://github.com/jupyter/notebook/issues/3704) env vars don't get loaded in when you run notebook cells in the GUI even if you're using the right kernel. I had no idea about that. But as long as I just run the notebook from the command line with:

```
jupyter nbconvert --to notebook --execute --inplace test-notebook.ipynb
```

They load just fine. So this part is all good.

### 5 - Importing PUDL libraries in a Jupyter notebook ✅
I am able to import libraries that should be available in `pudl-dev` into my notebook:

![image](https://github.com/user-attachments/assets/f568c850-8298-4569-a113-cf4dcc8c57a0)

### 6 - Loading other assets in a Python script ❌
If I try to execute this code in a Python script:

```
from dagster import AssetKey
from pudl.etl import defs
df = defs.load_asset_value(AssetKey("out_eia__yearly_plants"))
```

I get the `Database schema` error saying that I should run `alembic revision --autogenerate -m 'relevant message' && alembic upgrade head`. But that will fail due to the migration revision ID from step 2.

But for what it's worth, the underlying parquet file is there:

```
(pudl-dev) toposs-MBP:pudl sam$ ls -l $PUDL_OUTPUT/parquet/out_eia__yearly_plants.parquet
-rw-r--r--  1 sam  staff  2386780 Jun 11 09:27 /Users/sam/Documents/pudl-data/pudl_output/parquet/out_eia__yearly_plants.parquet
```

---

### Summary/Next steps
In summary, I'm wondering if I should just rerun the ETL again 🤷‍♂️. Thoughts?

**Reply by zaneselvans** [reply]
Okay, so it sounds like the previous problem was due to the environment variables not being set in the environment that your notebook was running in. But in your screenshot it looks like it's running inside `pudl-dev`.

How have you been setting your environment variables?  For whatever reason I'm not having this problem in VS Code. I set `$DAGSTER_HOME`, `$PUDL_INPUT` and `$PUDL_OUTPUT` in my `.zshrc`.  Maybe VS Code isn't using the same shell as you are at the terminal? What shell are you using when you open a MacOS terminal? What shell shows up in the VS Code terminal? Are they the same? Are `$DAGSTER_HOME` and `$PUDL_OUTPUT` set in both of those terminals?

Aside from the environment variable issue, it sounds like your database schema is out of sync with the one on `main`.   If you check out `main` and do

```console
alembic history
```

you'll see hashes & messages associated with the chain of database migrations it knows about.  If you do the same on your branch you should probably see some shared historical hash.  If you downgrade your DB to the last shared hash on your own branch, and then merge in `main` and use `alembic upgrade head` you should hopefully end up with a database that reflects the current schema on `main`, at which point you should be able to materialize assets locally again.

We have a little bit of [documentation on dealing with database migrations and Alembic](https://catalystcoop-pudl.readthedocs.io/en/nightly/dev/run_the_etl.html#database-schema-migration) but if it feels more confusing than its worth, you can also just remove `pudl.sqlite` entirely, and do `alembic upgrade head` to get an empty database with the right schema, and re-run the ETL.

**Reply by zaneselvans** [reply]
This command is only the right thing to do when what's happened is *you* have made some change to the schema locally on your branch, and you need to check in a migration that builds upon an existing version.

```
alembic revision --autogenerate -m 'relevant message' && alembic upgrade head
```

**Reply by zaneselvans** [reply]
It may also be faster and easier than running the ETL from scratch to just pull down the nightly build output and unzip it with something like:

```console
aws s3 cp s3://pudl.catalyst.coop/nightly/pudl.sqlite.zip $PUDL_OUTPUT
unzip $PUDL_OUTPUT/pudl.sqlite.zip -d $PUDL_OUTPUT
```

**Reply by seeess1** [reply]
In my case I'm not working on a feature branch. I'm just working with `main`, where I've fetched and merged in the latest changes from the origin.

On `main`, when I run `alembic history`, here's what I see:

```
49d2f4f7d7b7 -> 52aa9f44b6a7 (head), Add generator_operating_date to 860 changelog table
aee9c15c7394 -> 49d2f4f7d7b7, Update direct_support plant IDs to integers, create FK relationship between direct support IDs and harvested entities
<base> -> aee9c15c7394, Wipe and reset
```

Does that output indicate what I should do here?

And yep that makes sense about the `alembic revision --autogenerate -m 'relevant message' && alembic upgrade head` message. I'll ignore that for now since I'm just on `main` and not making any schema changes.

In the meantime, I will try that `aws` and `unzip` command that you shared to see if that gets me to a better spot.

**Reply by zaneselvans** [reply]
I suspect it's easiest to just download the DB from the nightly builds, which should match the alembic migrations on `main` and create then a new feature branch.

If you're just working on one branch where you need to change the schema, and the schema isn't changing on `main` at the same time this is pretty simple. As you add or change tables, you'll create migrations on your branch, and when you merge the branch into `main` they'll go along for the ride. It can also be nice to squash all of the accumulated DB migrations from your branch into a single migration before merging into `main` by downgrading to the last shared revision, removing the incremental migrations, and then creating a single new migration that contains all of the accumulated changes.  If you're switching between different branches that have different schema changes happening, or there are changes on `main` that you need to bring into your own branch it gets more complicated.

**Reply by seeess1** [reply]
### Assets/Dagster

After running the `aws` and `unzip` commands, I'm able to materialize the `core_ferc714__yearly_planning_area_demand_forecast` asset now, which is progress! But I also tried materializing another one (`_core_censusdp1tract__counties`) that currently shows up as failed when I load Dagster and it errored out with `sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) unable to open database file`.

### Environment/shell

On the environment/shell front, I did some more testing there. When I run a terminal in the VS Code GUI, it defaults to bash:

<img width="1433" alt="vs-output" src="https://github.com/user-attachments/assets/1eccc124-d7ae-4719-b20e-3603bd553c9c">

But notebooks are still failing to load assets as you can see in the screenshot above.

So I tried using `zsh` instead. But when I go to activate the environment (`mamba activate pudl-dev`), I get:

```
Run 'mamba init' to be able to run mamba activate/deactivate and start a new shell session. Or use conda to activate/deactivate.
```

<img width="1433" alt="vs-zsh" src="https://github.com/user-attachments/assets/c4f84ee9-cca8-45e5-8c3a-3e651343abeb">

Once I run `mamba init`, I get a list of files with no change and one line saying that `/Users/sam/.bash_profile` has been modified. At the end it says:

```
For changes to take effect, close and re-open your current shell.
```

But if I close the shell, open a new terminal, and then run `mamba activate pudl-dev`, I go right back to `Run 'mamba init' to be able to...` 😵‍💫.

At this point, I'm honestly okay with not being able to run notebooks. It would be nice if I could but I can load assets in regular Python scripts, which should be good enough for me. If I can just resolve the asset materialization issue that I noted at the top that should unblock me. Any thoughts on that one?

**Reply by zaneselvans** [reply]
Can you post the entire SQLite database error? There are several databases potentially in play here -- the Dagster has one that it uses to log events, there's the SQLite DB, and then there's a separate Census DP1 database, and I'm not sure which one the error is related to.

Do you have `censusdp1tract.sqlite` in your `$PUDL_OUTPUT` directory?

With the notebook, I think it'll be challenging to explore and debug a new asset without some way to interactively play with the outputs.

What is your login shell?  If you open a new terminal (not in VS Code) and run `echo $SHELL` what do you get? If you open a new terminal in VS Code, do you get the same thing?

**Reply by zaneselvans** [reply]
Rather than materializing a single asset, you might also try using `*censusdp1` in the search bar of the asset graph to select all assets upstream of those that contain the string `censusdp1` and then doing "Materialize All" -- if the problem is that you don't have the Census DP1 database locally that should regenerate it.

The docs on Dagster's [asset selection syntax](https://docs.dagster.io/concepts/assets/asset-selection-syntax) are really helpful.

**Reply by seeess1** [reply]
I'm good now on `_core_censusdp1tract__counties`! Looked up the related assets like you recommended and materialized everything in the lineage (just a group of four assets) and it all went through.  (The original error logs are attached. And before I ran the materialization I checked and confirmed that I had the `censusdp1tract.sqlite` file in my `$PUDL_OUTPUT` directory.)

Yeah I hear you on the complication with not being able to use a notebook. When I run `echo $SHELL` both in a regular terminal and in a a VS Code terminal, I get `/bin/bash`.

<img width="1433" alt="censusdp1" src="https://github.com/user-attachments/assets/c8dca7ee-cd60-4351-9fb9-11749ae0fe15">
